swarm.js

```diff
@@ -26,57 +39,100 @@ class Swarm extends EventEmitter {
         const info = peerInfo(null)
         info.connected(socket, isTCP)
         this.emit('connection', socket, info)
+        this.serverSockets += 1
+        this[kIncrPeerCount]()
+        socket.once('close', () => {
+          this.serverSockets -= 1
+          this.emit('disconnection', socket, info)
+          this[kDecrPeerCount]()
+        })
       },
       close: () => this.emit('close')
     })
-    queue.on('readable', this._drain(queue))
+
+    network.tcp.maxConnections = maxServerSockets
+    network.utp.maxConnections = maxServerSockets
+
+    queue.on('readable', this[kDrain](queue))
+
+    this.destroyed = false
+    this.clientSockets = 0
+    this.serverSockets = 0
     this.peers = 0
+
     this.maxPeers = maxPeers
-    this.emphemeral = ephemeral !== false
+    this.maxServerSockets = maxServerSockets
+    this.maxClientSockets = maxClientSockets
+
+    this.open = this.peers < this.maxPeers
+    this.ephemeral = ephemeral !== false
+
     this.network = network
-    this.queue = queue
+    this[kQueue] = queue
   }
-  _drain (queue) {
+  [kDrain] (queue) {
     const onConnect = (info) => (err, socket, isTCP) => {
       if (err) {
-        this.peers -= 1
+        this.clientSockets -= 1
+        this[kDecrPeerCount]()
         queue.requeue(info)
         drain()
         return
       }
       info.connected(socket, isTCP)
       this.emit('connection', socket, info)
       socket.on('close', () => {
-        this.peers -= 1
+        this.clientSockets -= 1
+        this.emit('disconnection', socket, info)
+        this[kDecrPeerCount]()
         info.disconnected()
         queue.requeue(info)
+        setImmediate(drain)
       })
       drain()
     }
     const drain = () => {
-      if (this.peers >= this.maxPeers) return
+      if (this.clientSockets >= this.maxClientSockets) return
       const info = queue.shift()
-      if (!info) return 
-      this.peers += 1
+      if (!info) return
+      this.clientSockets += 1
+      this[kIncrPeerCount]()
       this.network.connect(info.peer, onConnect(info))
     }
     return drain
   }
+  [kIncrPeerCount] () {
+    this.peers += 1
+    this.open = this.peers < this.maxPeers
+    if (this.open === false) {
+      this.network.tcp.maxConnections = -1
+      this.network.utp.maxConnections = -1
+    }
+  }
+  [kDecrPeerCount] () {
+    this.peers -= 1
+    if (this.open) return
+    this.open = this.peers < this.maxPeers
+    this.network.tcp.maxConnections = this.maxServerSockets
```

Comment:
only if open is true no?

Reply:
the only case where open would be false on init, is if maxPeers is 0
if that is the case maxConnections should be set to -1 to stop incoming conns

so the reason I've done done that is I can't see a case for it - and I feel like writing a test for it would be almost misleading. Happy to do it for the sake of symmetry if preferred though

Reply:
sorry got confused about where this was: essentially open is always true here because peers has been decremented